File: C++ solutions/1219. Path with Maximum Gold.cpp
class Solution {
public:
    void findMax(int i, int j, vector<vector<int>>& grid, int& m, int& n, int& ans, int& vp) {
        if (i < 0 || i >= n || j < 0 || j >= m || grid[i][j] == 0) return;
        else {
            int find = grid[i][j];
            vp += find;
            grid[i][j] = 0;
            ans = max(vp, ans);
            findMax(i + 1, j, grid, m, n, ans, vp);
            findMax(i - 1, j, grid, m, n, ans, vp);
            findMax(i, j + 1, grid, m, n, ans, vp);
            findMax(i, j - 1, grid, m, n, ans, vp);
            vp -= find;
            grid[i][j] = find;
        }
    }

    int getMaximumGold(vector<vector<int>>& grid) {
        int ans = 0;
        int vp = 0;
        int n = grid.size(), m = grid[0].size();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 0) {}
                else findMax(i, j, grid, m, n, ans, vp);
            }
        }
        return ans;
    }
};

File: Python solutions/1219. Path with Maximum Gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        def find_max(i, j, grid, m, n, ans, vp):
            if i < 0 or i >= n or j < 0 or j >= m or grid[i][j] == 0:
                return
            else:
                find = grid[i][j]
                vp[0] += find
                grid[i][j] = 0
                ans[0] = max(vp[0], ans[0])

                find_max(i + 1, j, grid, m, n, ans, vp)
                find_max(i - 1, j, grid, m, n, ans, vp)
                find_max(i, j + 1, grid, m, n, ans, vp)
                find_max(i, j - 1, grid, m, n, ans, vp)
                
                vp[0] -= find
                grid[i][j] = find
        
        ans = [0]
        vp = [0]
        n, m = len(grid), len(grid[0])
        for i in range(n):
            for j in range(m):
                if grid[i][j] != 0:
                    find_max(i, j, grid, m, n, ans, vp)

        return ans[0]
